README.md
# Headless ChatGPT

![Headless ChatGPT banner](/media/chatgpt-banner.webp)

## What is this

[Headless ChatGPT](https://www.npmjs.com/package/headless-chatgpt/) is a browser-based API emulator for ChatGPT. It is a local server that controls a `puppeteer` Chrome instance, which allows using this server like a ChatGPT API.

You do not need an API key. You just need a ChatGPT account.

In essence: I did not have API access because of a strange and persistent error, so I made my own API. Maybe it will be useful for someone else too.

## Software that use Headless ChatGPT:
- [CodeGPT Assistant](https://github.com/HalilCan/codeGPT-assistant-extension)
- [PR Assistant](https://www.youtube.com/watch?v=AgzSUK_9FoE)
- [Persona Generator](https://www.youtube.com/watch?v=xQByoQW1HjM)


## Usage

1. Start the server with `node <path-to-server.js> <preferred-port>`. This will start the puppet browser and the controller server will begin listening at the `<preferred-port>`.
2. Log in to ChatGPT and go through all pop-ups. 
3. You are ready to go.

Check the **API paths documentation** below for the API itself.

---

# API Paths Documentation
The server runs on a default port of 3000, unless specified otherwise when starting the server.

All you *really* need to get started with ChatGPT are `/queryAi` and `/retry`.

## GET /chatgpt
This endpoint navigates the browser to the ChatGPT page.

* Response: Success message or an error message if there was a problem navigating to the ChatGPT page.

## GET /start

This endpoint starts the browser.

* Response: Success message or an error message if there was a problem starting the browser.

## GET /visit
This endpoint visits a specified URL.

* Request: URL to visit as query parameter `url`
* Response: Success message containing the visited URL or an error message if there was a problem visiting the page.

## GET /close
This endpoint closes the browser.

* Response: Success message or an error message if there was a problem closing the browser.

## GET /currentChatList
This endpoint returns an array of the currently loaded chat names on the sidebar.

* Response: Either returns the list/array of past chats, or returns an error message.

## GET /currentGptList
This endpoint returns an array of the currently loaded gpt names on the sidebar, in addition to "GPT-4" and "GPT-3.5" even though they are hidden behind more clicks normally.

* Response: Either returns the list/array of recently used and available GPTs, or returns an error message.

## POST /type
This endpoint types a string into the browser.

* Request: JSON body with a string property string
* Response: Success message containing the typed string or an error message if there was a problem typing into the browser.

## POST /select
This endpoint selects an element in the browser using the provided selector.

*  Request: JSON body with a string property selector
* Response: JSON body with the boolean result of the operation or an error message if there was a problem selecting the element.

## POST /typeInElem
This endpoint types a string into the selected element in the browser.

* Request: JSON body with properties selector and string
* Response: JSON body with the boolean result of the operation or an error message if there was a problem typing into the selected element.

## POST /getInnerHtml
This endpoint gets the inner HTML of the selected element in the browser.

* Request: JSON body with a string property selector
* Response: JSON body with the innerHtml of the selected element or an error message if there was a problem retrieving the inner HTML.

## POST /getInnerHtmlOfLast
This endpoint gets the inner HTML of the last element that matches the provided selector in the browser.

* Request: JSON body with a string property selector
* Response: JSON body with the innerHtml of the selected element or an error message if there was a problem retrieving the inner HTML.

## POST /queryAi
This endpoint queries the AI with the provided text and optional context.

* Request: JSON body with a string property text and optional context
* Response: JSON body with the AI's response text or an error message if there was a problem querying the AI.

## POST /newChat
This endpoint starts a new chat with the desired model or gpt. Defaults to GPT-3.5.

* Request: JSON body with a "model" string field. This must contain the name of the GPT as it appears on the recent GPTs list, or it must contain one of "GPT-4" or "GPT-3.5". For backwards compatibility, sending "3" or "4" is also fine. Refer to `GET /currentGptList` for the exact strings.
* Response: Success message or error message.

## POST /selectChat
This endpoint selects and navigates to a previous chat the link to which is currently on the sidebar.

* Request: JSON body with a "chatName" string field. This must contain the name as it appears on the sidebar, or as has been received by the `GET /currentChatList` path. In case of conflicts, go with the latter.
* Response: Success message or error message.

## POST /loadMoreChats
This endpoint loads more past chats onto the sidebar.

* Request: Optional JSON body with a "isAllChats" field. If this field exists in the request, _all_ previous chats will be loaded onto the sidebar. This is a lengthy process and may lead to natural timeouts, depending on how many you have. A single update step without that is much faster in comparison.
* Response: Success message or error message.

## GET /retry
This endpoint retries the last query on the AI.

* Response: JSON body with the AI's response text or an error message if there was a problem retrying the last query.

---

## Installation instructions

1. Run `npm install headless-chatgpt`. 
1. Run `npm install` to install any missing dependencies.
2. Run `node node_modules/puppeteer/install.js` to make sure `puppeteer`'s puppet Chrome is successfully installed. This is a ~120MB download.

## How to contribute

Feel free to make pull requests or open issues in the [github repo](https://github.com/HalilCan/headless-chatgpt). A lot of new features can easily be added.


CHANGELOG.md
# Changelog

# 2.0
- 2.0.0:
    - The big fix: Openai changed the chatgpt interface sufficiently to break some existing functionality. This has been fixed.
    - New: Load and select older chats up to your first chat, if desired.
    - New: Select desired GPT model from the base models (4, 3.5, and Classic) in addition to whatever GPTs you have in your recent list.
    - New: Old cookies are now purged more cleanly, preventing the case where you are erroneously asked to log-in too many times - despite already logging in a previous session.

# 1.3
- 1.3.1: Chat type selection bug fixed.
- 1.3.0: Can start new chats with model selection.

## 1.2
- 1.2.2: Version mixup fixed.
- 1.2.0: Added cookie management. You should no longer have to log in repetedly (with new sessions, that is) as long as you submit at least one query per session.

## 1.1
- 1.1.0: Chatgpt broke. I unbroke it. (some changes in the interface require the browser puppet to change its behavior)

## 1.0
- 1.0.1: More informative readme and package.json with links.
- 1.0.0: Initial release

browser.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
const fs = require('fs');

puppeteer.use(StealthPlugin());

let browser;
let page;

const _regenButtonXPathSelector = "//div[@class='text-gray-400 flex self-end lg:self-center justify-center lg:justify-start mt-0 gap-1 visible']/button[2]" // as of today (5/12/2023) only the last response has a regen button. (The third child of the above div, being also a button) And thank Todd for that; he did it again.
const _continueButtonXPathSelector = "//button[contains(., 'Continue generating') or contains(., 'Continue')]";
const _newChatButtonXPathSelector = "(//a[@href="/" and contains(text(), chatgpt)])[2]";
//new chat button attempts: //a[@class='group flex h-10 items-center gap-2 rounded-lg px-2 font-medium hover:bg-token-surface-primary' and contains(., Chatgpt) and .//svg]

const _currentGptModeButtonSelector = "//div[@class='group flex cursor-pointer items-center gap-1 rounded-xl py-2 px-3 text-lg font-medium hover:bg-gray-50 radix-state-open:bg-gray-50 dark:hover:bg-black/10 dark:radix-state-open:bg-black/20']";
const _gptFourChatModeSelector = "(//div[@role='menuitem'])[1]";
const _gptThreeChatModeSelector = "(//div[@role='menuitem'])[2]";
const _pluginsChatModeSelector = "(//div[@role='menuitem'])[3]";
const _generalChatModeMenuButtonSelector = "//div[@role='menuitem']";

const _gptSelectorButtonStub = "//div[@class='pb-0.5 last:pb-0' and ";
const _gptSelectorButtons = "//div[@class='pb-0.5 last:pb-0']";
const _indexedGptSelectorButtonSelectorStub = "(//div[@class='pb-0.5 last:pb-0'])[";

const _oldChatButtonSelector = "//li[@class='relative']/div/a";
const _oldChatButtonSelectorTextStub = "//li[@class='relative']/div/a[contains(.,"; // add "STRING)]"
// working completed stub: //li[@class='relative']/div/a[contains(.,xpath)] no quotes at the end, strangely enough.

const _firstChatButtonSelector = "(//li[@class='relative']/div/a)[1]";
const _lastChatButtonSelector = "(//li[@class='relative']/div/a)[last()]" // this is for scrolling down so that chatgpt gives us older chats as well.

const _chatHistoryContainerSelector = "//nav[@aria-label='Chat history']"

// const _gptFourButtonXPathSelector = "//div[contains(., 'GPT-4')][contains(@class, 'group/button')]";
const _gptFourButtonXPathSelector = "//button[contains(., 'GPT-4')]";
const _gptThreeButtonXPathSelector = "//button[contains(., 'GPT-3.5')]";
const _maxWaitCount = 400;
const _generationWaitStepLength = 500;
const _generationInitialWaitLength = 500;
const _olderChatLoadTime = _generationInitialWaitLength * 6;

async function startBrowser(){
    browser = await puppeteer.launch({
        headless: false
    });
    page = await browser.newPage();
    await page.setViewport({
        width: 1366,
        height: 768,
        deviceScaleFactor: 1,
    });

    // getting saved cookies to avoid
    // Load cookies from JSON file
    try {
        let cookies = JSON.parse(fs.readFileSync('./cookies.json', 'utf8'));
        await page.setCookie(...cookies);
    } catch (e) {
    }

    // await page.emulate(iPad);
    // await page.emulateTimezone("Europe/Istanbul");
    await page.emulateTimezone("America/New_York");
}

async function visitPage(url){
    await page.goto(url);
}

async function type(string) {
    await page.keyboard.type(string);
}

async function closeBrowser(){
    await browser.close();
}

async function selectElem(selector, click=true) {
    const element = await page.waitForSelector(selector);
    //
    if (click) {
        await element.click();
    }
    return element;
}

// 0 indexed
async function selectElemWithIndex(selector, index) {    
    return page.$$eval(selector, elems => {
        if (elems.length >= index) {
            return elems[index];
        } else {
            if (elems.length <= 1) {
                return elems;
            } else {
                return elems[elems.length - 1];
            }
        }
    });
}

async function writeInTextArea(selector, string) {
    const element = await page.focus(selector);
    await type(string);
}

async function inputStringInSelector(selector, newValue) {
    await page.waitForSelector(selector);
    const element = await page.$(selector);
    if (element) {
        await page.$eval(selector, (el, value) => el.value = value, newValue);
    } else {
        console.error(`No element found with the selector ${selector}`);
    }
}

async function evalSelect(selector) {
    const result = await page.$$eval(selector, elems => {
        return elems;
    });
    return result;
}

async function evalSelectLastElem(selector) {
    const result = await page.$$eval(selector, elems => {
        const lastElem = elems.length > 1 ? elems[elems.length - 1] : elems[0];
        return lastElem;
    });
    return result;
}

// TODO: for the captcha, look for a button with inner text "Start Puzzle"


async function getInnerHtmlOfLastElem(selector) {
    const result = await page.$$eval(selector, elems => {
        const lastElem = elems.length > 1 ? elems[elems.length - 1] : elems[0];
        return lastElem.innerHTML;
    });
    return result;
}

async function getInnerHtml(selector) {
    const inner_html = await page.$eval(selector, element => element.innerHTML);
    return inner_html;
}

function timeout(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function clickButton(selector) {
    let buttons = await page.$x(selector);
    if (buttons.length > 0) {
        buttons[0].click();
        return 0;
    } else {
        return -1;
    }
}

async function readLastResponse() {
    const answerSelector = "div .markdown";
    let innerHTML = await getInnerHtmlOfLastElem(answerSelector);
    return innerHTML;
}

async function goToChat(chatName) {
    const chatSelector = _oldChatButtonSelectorTextStub + "\"" +  chatName + "\")]";
    clickResponse = clickButton(chatSelector);
    if (clickResponse === -1) {
        console.error(`Error: cannot click chat button for ${chatName}`);
        return clickResponse;
    }
    return 0;
}

async function getChatList () {
    const chatButtons = await page.$x("//li[@class='relative']/div/a");
    let list = [];
    for (const button of chatButtons) {
        const text = await page.evaluate(element => element.textContent, button);
        list.push(text.trim());
    }
    return list;
}

async function loadOlderChats(loadAllChats=false) {
    let chatCount = 0;
    let chatButtons = await page.$x("//li[@class='relative']/div/a");
    if (chatButtons.length < 3) {
        return;
    }
    
    if (loadAllChats) {
        let oldChatCount = chatCount;
        chatCount = chatButtons.length;
        while (oldChatCount != chatCount) {
            const secondLastElement = chatButtons[chatButtons.length - 2];
            await page.evaluate(element => {
                element.scrollIntoView({ behavior: 'smooth', block: 'end' });
            }, secondLastElement);

            await timeout(_olderChatLoadTime);
            chatButtons = await page.$x("//li[@class='relative']/div/a");     
            oldChatCount = chatCount;
            chatCount = chatButtons.length;
        }
    } else {
        const secondLastElement = chatButtons[chatButtons.length - 2];
        await page.evaluate(element => {
            element.scrollIntoView({ behavior: 'smooth', block: 'end' });
        }, secondLastElement);
        await timeout(_generationInitialWaitLength / 2);
    }
}

async function waitForNewChat() {
    let button = await page.$x(_currentGptModeButtonSelector);
    while (!button || button.length == 0) {
        await timeout(_generationInitialWaitLength);
        button = await page.$x(_currentGptModeButtonSelector);
    }
}

async function waitForGenerationToComplete() {
    // console.log("enter generation waiter");
    let button;
    let waitCount = 0;
    // sometimes our query doesn't "take" quickly, and the regenerate button from the last time around stays. This waits until that is done.
    button = await page.$x(_regenButtonXPathSelector);
    if (button && button.length > 0) {
        // console.log("first generation wait loop");
        while (button && button.length > 0) {
            // console.log("first generation inner loop, button:", button);
            await timeout(_generationInitialWaitLength);
            button = await page.$x(_regenButtonXPathSelector); 
        }
    }
    await timeout(_generationWaitStepLength / 2);

    // this just waits until the next regenerate button (which appears at the end of AI output) appears. If continue does so too, we press that and wait until done.
    while (waitCount < _maxWaitCount) {
        // console.log(`second generation wait loop : ${waitCount} / ${_maxWaitCount}`);
        await timeout(_generationWaitStepLength);
        button = await page.$x(_regenButtonXPathSelector);
        if (button.length > 0) {
            await timeout(_generationWaitStepLength);
            let continueButton = await page.$x(_continueButtonXPathSelector);
            if (continueButton.length > 0) {
                // TODO: Should this reset waitCount? Adverse behavior?
                waitCount = 0;
                continueButton[0].click();
            } else {
                break;
            }
        }
        waitCount++;
    }
}

async function getGptList() {
    let gptList = await page.$x(_gptSelectorButtons);
    let list = [];

    list.push("GPT-4");
    list.push("GPT-3.5");
    for (const gpt of gptList) {
        const text = await page.evaluate(element => element.textContent, gpt);
        if (text == "ChatGPTChatGPT") {
            continue;
        } else {
            list.push(text.trim());
        }
    }

    return list;
}

async function newChat(modelName) {
    // backwards compatibility:
    // console.log(`new chat model name: ${modelName}`);
    if (modelName == "4" || modelName == 4) {
        modelName = "GPT-4";
    }
    if (modelName == "3" || modelName == 3) {
        modelName = "GPT-3.5";
    }
    // console.log(`new chat modified model name: ${modelName}`);
    if (modelName == "GPT-4" || modelName == "GPT-3.5") {
        // go through the chatgpt button -> chatgpt selector -> appropriate gpt model
        // console.log(`new chat gpt 4 or 3.5 detected`);
        let clickResponse = await clickButton(_indexedGptSelectorButtonSelectorStub + "2]");
        if (clickResponse === -1) {
            console.error("Error: cannot click top chatgpt button for new chat");
            return clickResponse;
        }
        // console.log(`in new chatgpt`);
        await waitForNewChat();
        clickResponse = clickButton(_currentGptModeButtonSelector);
        if (clickResponse === -1) {
            console.error("Error: cannot click gpt selector dropdown menu");
            return clickResponse;
        }
        // console.log(`dropdown dropped`);
        await timeout(500);
        if (modelName == "GPT-4") {
            clickResponse = clickButton(_gptFourChatModeSelector);
            if (clickResponse === -1) {
                console.error("Error: cannot click gpt-4 button in the dropdown gpt selector menu");
                return clickResponse;
            }
        } else {
            clickResponse = clickButton(_gptThreeChatModeSelector);
            if (clickResponse === -1) {
                console.error("Error: cannot click gpt-4 button in the dropdown gpt selector menu");
                return clickResponse;
            }
        }
        // console.log(`selected from dropdown`);
    } else {
        // e.g. //div[@class='pb-0.5 last:pb-0' and @data-projection-id!='98' and contains(., "Hot Mods")] works
        const gptSelector = _gptSelectorButtonStub + `contains(., "${modelName}")]`;
        console.log(gptSelector);
        const clickResponse = await clickButton(gptSelector);
        if (clickResponse === -1) {
            console.error("Error: cannot click gpt button for new chat");
            return clickResponse;
        }
        await waitForNewChat();
        return 0;
    }
}

async function retry() {
    let innerHTML;
    const buttonXPathSelector = _regenButtonXPathSelector;
    const clickResponse = await clickButton(buttonXPathSelector);
    if (clickResponse === -1) {
        return clickResponse;
    }
    // TODO: I think in case of errors that need regeneration, they disable the input box. Wait and see.
    await waitForGenerationToComplete();
    innerHTML = await readLastResponse();
    return innerHTML;
}

async function queryAi(message, context) {
    let queryString, innerHTML;
    if (context === "") {
        queryString = message + "\n";
    } else {
        queryString = message + "\n WITH THE BELOW CONTEXT: \n" + context + "\n";
    }
    const inputSelector = "TextArea";
    const answerSelector = "div .markdown";

    await inputStringInSelector(inputSelector, queryString);
    // include the \n in the string because this doesn't work for some reason.
    // okay, that didn't work too - but using type for \n twice did.
    await writeInTextArea(inputSelector, "\n");
    await writeInTextArea(inputSelector, "\n");
    // console.log("the long wait beginn");
    
    await waitForGenerationToComplete();
    
    // console.log("the long wait ENDS");

    // const innerHTML = oldReturnMd || await getInnerHtmlOfLastElem(answerSelector);
    innerHTML += await getInnerHtmlOfLastElem(answerSelector);
    // console.log(innerHTML);
    return innerHTML;
}

async function saveCookies() {
    const cookies = await page.cookies();
    // Save cookies to a file
    try {
        fs.writeFileSync('./cookies.json', JSON.stringify(cookies, null, 2));
    } catch (e) {
        console.error(e);
    }
}


module.exports = { startBrowser, visitPage, closeBrowser, type, selectElem, 
    selectElemWithIndex, writeInTextArea, getInnerHtmlOfLastElem,
    queryAi, retry, saveCookies, newChat, loadOlderChats, getGptList, getChatList, goToChat};


server.js
const express = require('express');
const axios = require('axios');
const browserModule = require('./browser');
const bodyParser = require('body-parser');
const fs = require('fs');

const app = express();
app.use(bodyParser.json());

app.get('/chatgpt', async (req, res) => {
    try {
        await browserModule.startBrowser();
        await browserModule.visitPage("https://chat.openai.com/");
        res.send('Navigated to chatgpt');
    } catch (error) {
        console.error(error.message);
        res.status(500).send('Error navigating to chatgpt');
    }
});

app.get('/start', async (req, res) => {
    try {
        await browserModule.startBrowser();
        res.send('Browser started');
    } catch (error) {
        console.error(error);
        res.status(500).send('Error starting browser');
    }
});

app.get('/visit', async (req, res) => {
    try {
        await browserModule.visitPage(req.query.url);
        res.send(`Visited ${req.query.url}`);
    } catch (error) {
        console.error(error);
        res.status(500).send('Error visiting page');
    }
    if (!req.query.url) {
        res.send('Please provide a URL');
        return;
    }
});

app.get('/close', async (req, res) => {
    try {
        await browserModule.closeBrowser();
        res.send('Browser closed');
    } catch (error) {
        console.error(error);
        res.status(500).send('Error closing browser');
    }
});

app.get('/currentChatList', async (req, res) => {
    try {
        const list = await browserModule.getChatList();
        res.send(list);
    } catch (error) {
        console.error(error);
        res.status(500).send('Error getting the chat list');
    }
});

app.get('/currentGptList', async (req, res) => {
    try {
        const list = await browserModule.getGptList();
        res.send(list);
    } catch (error) {
        console.error(error);
        res.status(500).send('Error getting the gpt list');
    }
});

app.post('/loadMoreChats', async (req, res) => {
    try {
        if (!req.body.isAllChats) {
            await browserModule.loadOlderChats(false);
            res.send(`One page of older chats loaded`);
        } else {
            await browserModule.loadOlderChats(true);
            res.send(`All older chats loaded`);
        }
    } catch (error) {
        console.error(error);
        res.status(500).send('Error loading older chats');
    }
})

app.post('/type', async (req, res) => {
    try {
        if (!req.body.string) {
            res.send('Please provide a valid string to type.');
            return;
        }
        await browserModule.type(req.body.string);
        res.send(`Typed ${req.body.string}`);
    } catch (error) {
        console.error(error);
        res.status(500).send('Error typing in browser');
    }
})

app.post('/select', async (req, res) => {
    try {
        if (!req.body.selector) {
            res.send('Please provide a valid selector.');
            return;
        }
        const selected = await browserModule.selectElem(req.body.selector);
        res.send({"selected": selected});
    } catch (error) {
        console.error(error);
        res.status(500).send('Error selecting');
    }
})

app.post('/typeInElem', async (req, res) => {
    try {
        if (!req.body.selector || !req.body.string) {
            res.send('Please provide a valid selector and string.');
            return;
        }
        const selected = await browserModule.writeInTextArea(req.body.selector, req.body.string);
        res.send({"selected": selected});
    } catch (error) {
        console.error(error);
        res.status(500).send('Error typing in new selection');
    }
})

app.post('/getInnerHtml', async (req, res) => {
    try {
        if (!req.body.selector) {
            res.send('Please provide a valid selector.');
            return;
        }
        const innerHtml = await browserModule.getInnerHtml(req.body.selector);
        res.send({"innerHtml": innerHtml});
    } catch (error) {
        console.error(error);
        res.status(500).send('Error getting innerhtml');
    }
})

app.post('/getInnerHtmlOfLast', async (req, res) => {
    try {
        if (!req.body.selector) {
            res.send('Please provide a valid selector.');
            return;
        }
        const innerHtml = await browserModule.getInnerHtmlOfLastElem(req.body.selector);
        res.send({"innerHtml": innerHtml});
    } catch (error) {
        console.error(error);
        res.status(500).send('Error getting inner html of last elem from selector');
    }
})

app.post('/queryAi', async (req, res) => {
    try {
        if (!req.body.text) {
            res.send('Please provide valid text.');
            return;
        }
        let context = req.body.context ?? "";
        const innerHtml = await browserModule.queryAi(req.body.text, context);
        res.send({"text": innerHtml});
        // also save cookies since we've definitely logged in.
        await browserModule.saveCookies();
    } catch (error) {
        console.error(error);
        res.status(500).send('Error in querying AI');
    }
})

app.get('/retry', async (req, res) => {
    try {
        const innerHtml = await browserModule.retry();
        if (innerHtml === -1) {
            console.error(error);
            res.status(500).send('Error in retrying the last query on AI.');            
        }
        res.send({"text": innerHtml});
    } catch (error) {
        console.error(error);
        res.status(500).send('Error in retrying the last query on AI.');
    }
})

app.post('/selectChat', async (req, res) => {
    try {
        if (!req.body.chatName) {
            res.status(500).send(`Please include chatName in your request body`);
            return;
        }
        let chatName = req.body.chatName;
        // console.log(`model: ${model}`);
        const actionResponse = await browserModule.goToChat(chatName);
        if (actionResponse === -1) {
            console.error(`Error in selecting chat: ${chatName}`);
            res.status(500).send(`Error in selecting chat: ${chatName}`);            
        } else {
            res.send('Chat selected.');
        }
    } catch (error) {
        console.error(error);
        res.status(500).send(`Error in selecting existing chat.`);
    }
})

app.post('/newChat', async (req, res) => {
    try {
        let model;
        if (!req.body.model) {
            model = 3;
        } else {
            model = req.body.model;
        }
        // console.log(`model: ${model}`);
        const actionResponse = await browserModule.newChat(model);
        if (actionResponse === -1) {
            console.error('Error in starting new chat.');
            res.status(500).send('Error in starting new chat.');            
        } else {
            res.send('New chat started');
        }
    } catch (error) {
        console.error(error);
        res.status(500).send('Error in starting new chat.');
    }
})


// determine the port.
const defaultPort = 3000;
const port = process.argv[2] || defaultPort;

app.listen(port, async () => {
    console.log(`ChatGPT API server running on port ${port}`);
    const response = await axios.get(`http://localhost:${port}/chatgpt`);
    // console.log(response.data);
});